File: API/src/GoS.Domain/Resources/Models/HeroAbilities/HeroAbility.cs
using System.Text.Json.Serialization;

namespace GoS.Domain.Resources.Models.HeroAbilities;

public class HeroAbility
{
    [JsonPropertyName("abilities")]
    public List<string> Abilities { get; set; } = [];

    [JsonPropertyName("talents")]
    public List<Talent> Talents { get; set; } = [];

    [JsonPropertyName("facets")]
    public List<Facet> Facets { get; set; } = [];

    [JsonPropertyName("shard")]
    public required Shard Shard { get; set; }

    [JsonPropertyName("aghanim_scepter")]
    public required AghanimScepter AghanimScepter { get; set; }
}

public class AghanimScepter
{
    [JsonPropertyName("scepter_desc")]
    public string ScepterDesc { get; set; }

    [JsonPropertyName("scepter_skill_name")]
    public string ScepterSkillName { get; set; }

    [JsonPropertyName("video")]
    public string Video { get; set; }
}

public class Shard
{
    [JsonPropertyName("shard_desc")]
    public string ShardDesc { get; set; }

    [JsonPropertyName("shard_skill_name")]
    public string ShardSkillName { get; set; }

    [JsonPropertyName("video")]
    public string Video { get; set; }
}

File: gos-client/src/types/heroAbility.ts
import type {BooleanState} from "./common.ts";

export interface Facet {
    id: number;
    name: string; // internal name
    icon: string;
    color: string; // "Red", "Blue", "Gray", etc.
    title: string;
    description: string;
    deprecated: BooleanState | null;
    abilities: string[] | null;
}

export interface Talent {
    name: string;
    level: TalentLevel;
}

export type TalentLevel = 1 | 2 | 3 | 4;

export interface Shard {
  shard_desc: string;
  shard_skill_name: string;
  video: string;
}

export interface AghanimScepter {
  scepter_desc: string;
  scepter_skill_name: string;
  video: string;
}

export interface HeroAbility {
  abilities: string[];
  talents: Talent[];
  facets: Facet[];
  shard: Shard;
  aghanim_scepter: AghanimScepter;
}
